File: benchmarks/solutions/zig/s1-channel-pipeline.zig
const std = @import("std");

fn Channel(comptime T: type) type {
    return struct {
        buf: [32]T = undefined,
        head: usize = 0,
        tail: usize = 0,
        count: usize = 0,
        closed: bool = false,
        mutex: std.Thread.Mutex = .{},
        not_empty: std.Thread.Condition = .{},
        not_full: std.Thread.Condition = .{},

        const Self = @This();

        fn send(self: *Self, val: T) void {
            self.mutex.lock();
            defer self.mutex.unlock();
            while (self.count == 32) self.not_full.wait(&self.mutex);
            self.buf[self.tail] = val;
            self.tail = (self.tail + 1) % 32;
            self.count += 1;
            self.not_empty.signal();
        }

        fn recv(self: *Self) ?T {
            self.mutex.lock();
            defer self.mutex.unlock();
            while (self.count == 0 and !self.closed) self.not_empty.wait(&self.mutex);
            if (self.count == 0) return null;
            const val = self.buf[self.head];
            self.head = (self.head + 1) % 32;
            self.count -= 1;
            self.not_full.signal();
            return val;
        }

        fn close(self: *Self) void {
            self.mutex.lock();
            defer self.mutex.unlock();
            self.closed = true;
            self.not_empty.broadcast();
        }
    };
}

pub fn main() !void {
    var ch1 = Channel(i32){};
    var ch2 = Channel(i32){};

    const t1 = try std.Thread.spawn(.{}, struct {
        fn run(ch: *Channel(i32)) void {
            for (1..21) |i| ch.send(@intCast(i));
            ch.close();
        }
    }.run, .{&ch1});

    const t2 = try std.Thread.spawn(.{}, struct {
        fn run(in_ch: *Channel(i32), out_ch: *Channel(i32)) void {
            while (in_ch.recv()) |v| {
                if (@mod(v, 2) == 1) out_ch.send(v);
            }
            out_ch.close();
        }
    }.run, .{ &ch1, &ch2 });

    const stdout = std.io.getStdOut().writer();
    while (ch2.recv()) |v| {
        try stdout.print("{}\n", .{v});
    }

    t1.join();
    t2.join();
}

File: benchmarks/solutions/cpp/s1-channel-pipeline.cpp
#include <iostream>
#include <thread>
#include <queue>
#include <mutex>
#include <condition_variable>
#include <optional>

template<typename T>
class Channel {
    std::queue<T> buf;
    std::mutex mtx;
    std::condition_variable cv;
    bool closed = false;
public:
    void send(T val) {
        std::lock_guard lk(mtx);
        buf.push(std::move(val));
        cv.notify_one();
    }
    void close() {
        std::lock_guard lk(mtx);
        closed = true;
        cv.notify_all();
    }
    std::optional<T> recv() {
        std::unique_lock lk(mtx);
        cv.wait(lk, [&] { return !buf.empty() || closed; });
        if (buf.empty()) return std::nullopt;
        T val = std::move(buf.front());
        buf.pop();
        return val;
    }
};

int main() {
    Channel<int> ch1, ch2;

    std::thread producer([&] {
        for (int i = 1; i <= 20; i++) ch1.send(i);
        ch1.close();
    });

    std::thread filter([&] {
        while (auto v = ch1.recv()) {
            if (*v % 2 == 1) ch2.send(*v);
        }
        ch2.close();
    });

    while (auto v = ch2.recv()) {
        std::cout << *v << "\n";
    }

    producer.join();
    filter.join();
    return 0;
}
